crypto/utils/utils.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 Nikos Filippakis
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package utils
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/base64"
+	"math/rand"
+	"strings"
+
+	"github.com/btcsuite/btcutil/base58"
+	"golang.org/x/crypto/hkdf"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+const (
+	// AESCTRKeyLength is the length of the AES256-CTR key used.
+	AESCTRKeyLength = 32
+	// AESCTRIVLength is the length of the AES256-CTR IV used.
+	AESCTRIVLength = 16
+	// HMACKeyLength is the length of the HMAC key used.
+	HMACKeyLength = 32
+	// SHAHashLength is the length of the SHA hash used.
+	SHAHashLength = 32
+)
+
+// XorA256CTR encrypts the input with the keystream generated by the AES256-CTR algorithm with the given arguments.
+func XorA256CTR(source []byte, key [AESCTRKeyLength]byte, iv [AESCTRIVLength]byte) []byte {
+	block, _ := aes.NewCipher(key[:])
+	result := make([]byte, len(source))
+	cipher.NewCTR(block, iv[:]).XORKeyStream(result, source)
+	return result
+}
+
+// GenAttachmentA256CTR generates a new random AES256-CTR key and IV suitable for encrypting attachments.
+func GenAttachmentA256CTR() (key [AESCTRKeyLength]byte, iv [AESCTRIVLength]byte) {
+	_, err := rand.Read(key[:])
+	if err != nil {
+		panic(err)
+	}
+
+	// For some reason we leave the 8 last bytes empty even though AES256-CTR has a 16-byte block size.
+	_, err = rand.Read(iv[:8])
+	if err != nil {
+		panic(err)
+	}
+	return
+}
+
+// GenA256CTRIV generates a random IV for AES256-CTR with the last bit set to zero.
+func GenA256CTRIV() (iv [AESCTRIVLength]byte) {
+	_, err := rand.Read(iv[:])
+	if err != nil {
+		panic(err)
+	}
+	iv[8] &= 0x7F
```

Comment:
Shouldn't this modify index 7 (8th byte) and use `0x7E` instead of `0x7F` to set bit 63 to 0?

```suggestion
	iv[7] &= 0b11111110
```

(The master branch has the same code for making the IV for key exports)

Reply:
I wasn't sure myself, I wrote this judging by https://github.com/matrix-org/matrix-js-sdk/blob/048d7a9b6377c155c56d5981e9b5d32e060398d2/src/crypto/aes.js#L50

Reply:
I'm not sure either, but I thought `iv[8] &= 0x7F` sets bit 64 (65th bit), while `iv[7] &= 0xFE` sets bit 63 (64th bit). Maybe we should ask someone who actually knows how bits and byte arrays work

Reply:
To add to the confusion, Element Android [seems to do](https://github.com/vector-im/element-android/blob/develop/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/secrets/DefaultSharedSecretStorageService.kt#L269-L272) `iv[9] &= 0x7F`?!

Reply:
Apparently bits are weird and Android/iOS are doing it wrong, but the js-sdk way is correct